cmd/snap/cmd_download.go

```diff
@@ -170,3 +159,141 @@ func (x *cmdDownload) Execute(args []string) error {
+		return err
+	}
+	f, err := os.Create(downloadPath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	if _, err := io.Copy(f, stream); err != nil {
+		return err
+	}
+
+	fmt.Fprintf(Stdout, i18n.G("Fetching assertions for %q\n"), snapName)
+
+	assertFname := strings.TrimSuffix(downloadPath, filepath.Ext(downloadPath)) + ".assert"
+	sha3_384, _, err := asserts.SnapFileSHA3_384(downloadPath)
+	if err != nil {
+		return err
+	}
+
+	db, err := asserts.OpenDatabase(&asserts.DatabaseConfig{
+		Backstore: asserts.NewMemoryBackstore(),
+		Trusted:   sysdb.Trusted(),
+	})
+	if err != nil {
+		return err
+	}
+	retrieve := func(ref *asserts.Ref) (asserts.Assertion, error) {
+		headers, err := asserts.HeadersFromPrimaryKey(ref.Type, ref.PrimaryKey)
+		if err != nil {
+			return nil, err
+		}
+		asserts, err := x.client.Known(ref.Type.Name, headers, &client.KnownOptions{Remote: true})
+		if err != nil {
+			return nil, err
+		}
+		if len(asserts) != 1 {
+			return nil, fmt.Errorf("unexpected amount of assertions: %q", asserts)
+		}
+		return asserts[0], nil
+	}
+
+	w, err := os.Create(assertFname)
+	if err != nil {
+		return err
+	}
+	defer w.Close()
+
+	encoder := asserts.NewEncoder(w)
+	save := func(a asserts.Assertion) error {
+		encoder.Encode(a)
+		return db.Add(a)
+	}
+	ref := &asserts.Ref{asserts.SnapRevisionType, []string{sha3_384}}
+	fetcher := asserts.NewFetcher(db, retrieve, save)
+	if err := fetcher.Fetch(ref); err != nil {
+		return err
+	}
+	// XXX: cross check?
+
+	return nil
+}
+
+func isErrorKindLoginRequired(e error) bool {
+	if err, ok := e.(*client.Error); ok {
+		return err.Kind == client.ErrorKindLoginRequired
+	}
+	return false
+}
+
+func isConnectionError(err error) bool {
+	var connErr client.ConnectionError
+	return xerrors.As(err, &connErr)
+}
+
+func (x *cmdDownload) Execute(args []string) error {
+	if strings.ContainsRune(x.Basename, filepath.Separator) {
+		return fmt.Errorf(i18n.G("cannot specify a path in basename (use --target-dir for that)"))
+	}
+	if err := x.setChannelFromCommandline(); err != nil {
+		return err
+	}
+
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
+	// "snap download" works only in --direct mode on non-linux
+	if runtime.GOOS != "linux" {
+		x.Direct = true
+	}
+
+	var revision snap.Revision
+	if x.Revision == "" {
+		revision = snap.R(0)
+	} else {
+		if x.Channel != "" {
+			return fmt.Errorf(i18n.G("cannot specify both channel and revision"))
+		}
+		if x.CohortKey != "" {
+			return fmt.Errorf(i18n.G("cannot specify both cohort and revision"))
+		}
+		var err error
+		revision, err = snap.ParseRevision(x.Revision)
+		if err != nil {
+			return err
+		}
+	}
+	snapName := string(x.Positional.Snap)
+
+	if x.Direct {
+		return x.downloadDirect(snapName, revision)
+	}
+
+	err := x.downloadViaSnapd(snapName, revision)
+	if isConnectionError(err) || isErrorKindLoginRequired(err) {
+		fmt.Fprintf(Stdout, "Cannot connect to the daemon, trying direct download\n")
```

Comment:
do we really want to print this? should it go to Stderr?  even if we do, it should be probably be prefix, as some thing in `snap info`,  also missing i18n ?